Add graphthewy/1.1 - Header-only graph modelling library

alex-87 opened the pull request:
**graphthewy/1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=4814) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=4814) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=4814) it.</sub>

uilianries commented on recipes/graphthewy/all/conanfile.py:
```diff
@@ -0,0 +1,51 @@
+
+class GraphthewyConan(ConanFile):
+    name = "graphthewy"
+    version = "1.1"
```
```suggestion
```

Version should be solved by command line, e.g.  conan create . graphthewy/1.1@

alex-87 replied:
Thank you for the review. PR updated.

prince-chrismc commented on recipes/graphthewy/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,22 @@
+
+    def test(self):
+        if not tools.cross_building(self):
+            os.chdir("bin")
+            self.run(".%stest_package" % os.sep)
```
```suggestion
            bin_path = os.path.join("bin", "test_package")
            self.run(bin_path, run_environment=True)
```

prince-chrismc reviewed (commented):
Some very minor comments

alex-87 commented:
Thank you very much for the review!

In accordance with @prince-chrismc's review suggestions, I have :

- [x] Fixed the `test_package/conanfile.py`
- [x] Removed the `check_min_cppstd(self, 17)` call as it is handled by recipe
- [x] Cleaned the package's test source
- [x] Removed the license of the test's source

prince-chrismc commented:
> * Removed the `check_min_cppstd(self, 17)` call as it is handled by recipe

You removed the wrong one by accident!

alex-87 commented:
Thank you for the explanation, I've done the following modifications: 

- To remove `check_min_std(self, 17)` call from `test_package/conanfile.py` as this does not need any extra processing,
- To add the `check_min_std(self, 17)` call to the recipe `graphthewy/all/conanfile.py` as just before previous change

SpaceIm commented on recipes/graphthewy/all/conanfile.py:
```diff
@@ -0,0 +1,50 @@
+        if self.settings.compiler.cppstd:
+            tools.check_min_cppstd(self, 17)
+
+        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
+        if minimum_version:
+            if tools.Version(self.settings.compiler.version) < minimum_version:
+                raise ConanInvalidConfiguration("graphthewy requires C++17, which your compiler does not support.")
+        else:
+            self.output.warn("graphthewy requires C++17. Your compiler is unknown. Assuming it supports C++17.")
```
```suggestion

        def lazy_lt_semver(v1, v2):
            lv1 = [int(v) for v in v1.split(".")]
            lv2 = [int(v) for v in v2.split(".")]
            min_length = min(len(lv1), len(lv2))
            return lv1[:min_length] < lv2[:min_length]

        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
        if not minimum_version:
            self.output.warn("graphthewy requires C++17. Your compiler is unknown. Assuming it supports C++17.")
        elif lazy_lt_semver(str(self.settings.compiler.version), minimum_version):
            raise ConanInvalidConfiguration("graphthewy requires C++17, which your compiler does not support.")
```

Trick to not raise if `compiler.version = 15` in profile for Visual Studio

SpaceIm commented on recipes/graphthewy/all/conanfile.py:
```diff
@@ -0,0 +1,50 @@
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("*.hpp", "include/graphthewy", keep_path=False)
```
```suggestion
        self.copy("*.hpp", dst=os.path.join("include", "graphthewy"), src=os.path.join("src", "include"))
```

SpaceIm commented on recipes/graphthewy/all/conanfile.py:
```diff
@@ -0,0 +1,50 @@
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "Simple header-only C++ Library for graph modelling (directed or not) and graph cycle detection. "
+    topics = ("graph", "algorithm", "modelling", "header-only")
+    settings = "os", "compiler"
```
```suggestion
    settings = "compiler"
```

SpaceIm commented on recipes/graphthewy/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+add_executable(test_package test_package.cpp)
```
```suggestion
add_executable(test_package test_package.cpp)
target_link_libraries(test_package ${CONAN_LIBS})
```

It works without due to conan_basic_setup() and because your lib is header-only, but it's consistent with others recipes.

alex-87 replied:
Thank you for the explanation, however it does not build locally without setting `-s compiler.cppstd=17` to `conan create` command. Am I missing a requirement, especially on the recipe ?

SpaceIm replied:
oh yes, you must also add `set_property(TARGET test_package PROPERTY CXX_STANDARD 17)`

SpaceIm commented on recipes/graphthewy/all/conanfile.py:
```diff
@@ -0,0 +1,55 @@
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("*.hpp", dst=os.path.join("include", "graphthewy"), src=self._source_subfolder, keep_path=False)
```
ok let's keep this, actually I missed self._source_subfolder

so you would like `src=os.path.join(self._source_subfolder, "src", "include")`and remove keep_path, so that there is no chance you include a .hpp you don't want from another location.
